File: go/problem_9.go
package main

import "fmt"

// largestSumNonAdjacent returns the largest sum of non-adjacent elements.
// Uses O(1) space DP: at each step, either skip the current element (keep
// prev) or take it (prevprev + current). This runs in O(n) time.
func largestSumNonAdjacent(nums []int) int {
	prevPrev, prev := 0, 0
	for _, n := range nums {
		curr := prev
		if prevPrev+n > curr {
			curr = prevPrev + n // take current element
		}
		prevPrev, prev = prev, curr
	}
	return prev
}

func problem9() {
	fmt.Println("Problem 9:")
	if largestSumNonAdjacent([]int{2, 4, 6, 2, 5}) != 13 { panic("assertion failed") }
	if largestSumNonAdjacent([]int{5, 1, 1, 5})    != 10 { panic("assertion failed") }
	if largestSumNonAdjacent([]int{})               != 0  { panic("assertion failed") }
	if largestSumNonAdjacent([]int{3})              != 3  { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: c/problem_9.c
#include "problem_9.h"
#include <stdio.h>
#include <assert.h>

/* largestSumNonAdj: O(1) space DP. */
long long largestSumNonAdj(const int *nums, int n) {
    long long prevPrev = 0, prev = 0;
    for (int i = 0; i < n; i++) {
        long long curr = prev > prevPrev + nums[i] ? prev : prevPrev + nums[i];
        prevPrev = prev; prev = curr;
    }
    return prev;
}

void problem9(void) {
    printf("Problem 9:\n");
    int a[] = {2,4,6,2,5}, b[] = {5,1,1,5}, c[] = {3};
    assert(largestSumNonAdj(a, 5) == 13);
    assert(largestSumNonAdj(b, 4) == 10);
    assert(largestSumNonAdj(NULL, 0) == 0);
    assert(largestSumNonAdj(c, 1) == 3);
    printf("All Tests Passed\n");
}
